public/compute/the_hairball.md

```diff
@@ -1,10 +1,174 @@
+Each parameter may only be specified in one file.
+The resulting list of parameters will be created by taking all possible combinations of parameter combinations from each file.
+
+#### Simple example
+Say we have the following parameter combinations:
+
+|input|
+|-----|
+|hello|
+|bye  |
+
+and
+
+|workflowName   |creationDate|
+|---------------|------------|
+|myFirstWorkflow|today       |
+
+Then these will get combined to 
+
+|input|workflowName   |creationDate|
+|-----|---------------|------------|
+|hello|myFirstWorkflow|today       |
+|bye  |myFirstWorkflow|today       |
+
+#### Example with three files
+
+Say we have the following three files:
+
+|input|
+|-----|
+|hello|
+|bye  |
+
+|sample |
+|-------|
+|sample1|
+|sample2|
+
+|workflowName   |creationDate|
+|---------------|------------|
+|myFirstWorkflow|today       |
+
+Then these will get combined to 
+
+|input|sample |workflowName   |creationDate|
+|-----|-------|---------------|------------|
+|hello|sample1|myFirstWorkflow|today       |
+|hello|sample2|myFirstWorkflow|today       |
+|bye  |sample1|myFirstWorkflow|today       |
+|bye  |sample2|myFirstWorkflow|today       |
+
+#### Collapsing (or Folding) the table of Parameter values
+Say we have the following combinations:
+
+|project |dir |sample |
+|--------|----|-------|
+|project1|dir1|sample1|
+|project1|dir2|sample2|
+|project1|dir2|sample3|
+|project2|dir2|sample4|
+
+Now if a particular step's protocol depends only on the value of parameter `project`, it only needs to be run twice, once for `project=project1` and once for `project=project2`.
+So the table of parameter combinations can be collapsed to the much smaller table.
+
+|project |
+|--------|
+|project1|
+|project2|
+
+If the step's protocol depends on parameters `project` and `dir`, three instances for the step need to be created, collapsing the table as follows:
+
+|project |dir |
+|--------|----|
+|project1|dir1|
+|project1|dir2|
+|project2|dir2|
+
+The step need not be run for the combination `project=project1, dir=dir2`, since no parameter combinations exits with those values.
+
+#### environment files
+Data is shared between steps using env files.
+Those contain entries of the shape `name[index]=value`
+
+At generation time, compute generates `user.env` which contains all inputs for all steps that are known at generation time.
+
+At runtime, the steps may create files named *stepname*_*stepindex*.env. So for example `step1_0.sh` creates `step1_0.env`.
+
 #### Weaving
-Parameters, which are known before hand can be connected to the environment file or weaved directly in the protocols (if 'weave' flag is set in command-line options). In our example, two shell scripts are generated for 
+Parameters which are known beforehand can also be **weaved** directly into the protocols (if 'weave' flag is set in command-line options). In our example, two shell scripts are generated for 
 the 'step1'. The weaved version of generated files are shown below.
```

Comment:
In our example, two shell scripts are generated for  'step1'. The weaved version of the generated files are shown below.